clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -502,16 +502,26 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
       // TODO(cir): CGFPOptionsRAII
       assert(!MissingFeatures::CGFPOptionsRAII());
 
-      if (type->isHalfType() && !CGF.getContext().getLangOpts().NativeHalfType)
-        llvm_unreachable("__fp16 type NYI");
+      if (type->isHalfType() &&
+          !CGF.getContext().getLangOpts().NativeHalfType) {
+        // Another special case: half FP increment should be done via float
+        if (CGF.getContext().getTargetInfo().useFP16ConversionIntrinsics()) {
+          value = Builder.createCastFromFP16(CGF.getLoc(E->getExprLoc()), input,
+                                             CGF.CGM.FloatTy);
+        } else {
+          value = Builder.createCast(CGF.getLoc(E->getExprLoc()),
+                                     mlir::cir::CastKind::floating, input,
+                                     CGF.CGM.FloatTy);
```

Comment:
Do we have a real world case for this else branch to test?
 I'm thinking about how we lower this CastOp in current LLVM Lowering. 
it looks like it may exercise this lowering code 
[`rewriter.replaceOpWithNewOp<mlir::LLVM::FPExtOp>`](
https://github.com/llvm/clangir/blob/24b24557c98d1c031572a567b658cfb6254f8a89/clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp#L769)
And [mlir::LLVM::FPExtOp is expecting floating type arg](https://mlir.llvm.org/docs/Dialects/LLVM/#llvmfpext-llvmfpextop),
But your type converter converts   cir.f16.storage into MLIR::IntegerType.....

Reply:
This else branch is the normal case and it is hit when `input` is of type `!cir.f16`. When `input` is of type `!cir.f16.storage`, the then branch would be taken instead.

> Do we have a real world case for this else branch to test?

Well since the else branch is the normal one, let's consider instead the then branch. I searched in the upstream clang tests with `llvm.convert.from.fp16` and I found nothing. For most targets the `useFP16ConversionIntrinsics()` function call would return `false`, which indicates for most targets the else branch would be taken. But there are indeed some exceptions, for example `mips-unknown-linux-gnu`: https://godbolt.org/z/feGvPYGxE

Anyway for now I think I'm going to remove the `!cir.f16.storage` stuff and make it assert on relevant paths, including the then branch here.

Reply:
Yep, you are right, the else branch is the most common case, and it should takes existing cir.f16 type as input. And lowering supports it without problem.